File: resource-server/src/lib/vault.ts
import fs from "fs";
import path from "path";

/**
 * Resolves and reads a single .md note from the vault.
 * section = subfolder path (e.g. "Blockchain/00 - Overview")
 * file    = filename without .md (e.g. "NZRT Blockchain Architecture")
 */
export function readNote(section: string, file: string): string {
  const VAULT_PATH = process.env.VAULT_PATH || "C:\\Users\\Natha\\Nextcloud2\\Documents\\Obsidian";
  const filePath = path.resolve(VAULT_PATH, section, `${file}.md`);

  // Prevent path traversal outside vault
  if (!filePath.startsWith(path.resolve(VAULT_PATH))) {
    throw new Error("Path traversal not allowed");
  }

  if (!fs.existsSync(filePath)) {
    throw new Error(`Note not found: ${section}/${file}`);
  }

  return fs.readFileSync(filePath, "utf-8");
}

/**
 * Simple grep-style search across all .md files in vault.
 * Returns up to 10 matches: { file, excerpt }.
 */
/** Strip YAML frontmatter and inline tag lines from content */
function stripFrontmatter(content: string): string {
  // Split on --- lines (handles CRLF + LF + BOM)
  const parts = content.replace(/^﻿/, "").split(/^-{3}\s*$/m);
  const body = parts.length >= 3 ? parts.slice(2).join("---") : content;
  // Strip [#"tag"] inline lines
  return body.replace(/^(\[#[^\]]+\]\s*)+\r?\n?/gm, "").trim();
}

export function searchVault(query: string): { file: string; excerpt: string }[] {
  const VAULT_PATH = process.env.VAULT_PATH || "C:\\Users\\Natha\\Nextcloud2\\Documents\\Obsidian";
  const results: { file: string; excerpt: string }[] = [];
  const lowerQuery = query.toLowerCase();

  function walk(dir: string): void {
    if (results.length >= 10) return;
    const entries = fs.readdirSync(dir, { withFileTypes: true });
    for (const entry of entries) {
      if (results.length >= 10) break;
      const fullPath = path.join(dir, entry.name);
      if (entry.isDirectory()) {
        walk(fullPath);
      } else if (entry.name.endsWith(".md")) {
        const raw = fs.readFileSync(fullPath, "utf-8");
        const content = stripFrontmatter(raw);
        const idx = content.toLowerCase().indexOf(lowerQuery);
        if (idx !== -1) {
          const start = Math.max(0, idx - 80);
          const end = Math.min(content.length, idx + 300);
          const relative = path.relative(VAULT_PATH, fullPath);
          results.push({ file: relative, excerpt: content.slice(start, end).trim() });
        }
      }
    }
  }

  walk(VAULT_PATH);
  return results;
}

File: python-resource-server/vault.py
from __future__ import annotations
import os
import re

VAULT_PATH = os.getenv("VAULT_PATH", "/home/nzrtnetw/vault")


def strip_frontmatter(content: str) -> str:
    content = content.lstrip("﻿")
    parts = re.split(r"^---\s*$", content, flags=re.MULTILINE)
    body = "".join(parts[2:]) if len(parts) >= 3 else content
    return re.sub(r"^(\[#[^\]]+\]\s*)+\r?\n?", "", body, flags=re.MULTILINE).strip()


def read_note(section: str, file: str) -> str:
    vault = os.path.realpath(VAULT_PATH)
    target = os.path.realpath(os.path.join(vault, section, f"{file}.md"))
    if not target.startswith(vault + os.sep) and target != vault:
        raise ValueError("Path traversal not allowed")
    if not os.path.isfile(target):
        raise FileNotFoundError(f"Note not found: {section}/{file}")
    with open(target, encoding="utf-8") as f:
        return f.read()


def search_vault(query: str, subfolder: str = "") -> list[dict]:
    if subfolder:
        search_root = os.path.realpath(os.path.join(VAULT_PATH, subfolder))
        vault = os.path.realpath(VAULT_PATH)
        if not search_root.startswith(vault + os.sep):
            raise ValueError("Path traversal not allowed")
    else:
        search_root = VAULT_PATH
    results: list[dict] = []
    lower = query.lower()
    for root, _, files in os.walk(search_root):
        if len(results) >= 10:
            break
        for name in files:
            if not name.endswith(".md") or len(results) >= 10:
                continue
            path = os.path.join(root, name)
            try:
                with open(path, encoding="utf-8") as f:
                    raw = f.read()
            except OSError:
                continue
            content = strip_frontmatter(raw)
            idx = content.lower().find(lower)
            if idx == -1:
                continue
            start = max(0, idx - 80)
            end = min(len(content), idx + 300)
            rel = os.path.relpath(path, VAULT_PATH)
            results.append({"file": rel, "excerpt": content[start:end].strip()})
    return results
